mne/morph.py

```diff
@@ -0,0 +1,934 @@
+        and :ref:`Logging documentation <tut_logging>` for more).
+
+    Returns
+    -------
+    morph_matrix : sparse matrix
+        matrix that morphs data from ``subject_from`` to ``subject_to``.
+
+    Notes
+    -----
+    This function can be used to morph data between hemispheres by setting
+    ``xhemi=True``. The full cross-hemisphere morph matrix maps left to right
+    and right to left. A matrix for cross-mapping only one hemisphere can be
+    constructed by specifying the appropriate vertices, for example, to map the
+    right hemisphere to the left:
+    ``vertices_from=[[], vert_rh], vertices_to=[vert_lh, []]``.
+
+    Cross-hemisphere mapping requires appropriate ``sphere.left_right``
+    morph-maps in the subject's directory. These morph maps are included
+    with the ``fsaverage_sym`` FreeSurfer subject, and can be created for other
+    subjects with the ``mris_left_right_register`` FreeSurfer command. The
+    ``fsaverage_sym`` subject is included with FreeSurfer > 5.1 and can be
+    obtained as described `here
+    <http://surfer.nmr.mgh.harvard.edu/fswiki/Xhemi>`_. For statistical
+    comparisons between hemispheres, use of the symmetric ``fsaverage_sym``
+    model is recommended to minimize bias [1]_.
+
+    References
+    ----------
+    .. [1] Greve D. N., Van der Haegen L., Cai Q., Stufflebeam S., Sabuncu M.
+           R., Fischl B., Brysbaert M.
+           A Surface-based Analysis of Language Lateralization and Cortical
+           Asymmetry. Journal of Cognitive Neuroscience 25(9), 1477-1492, 2013.
+    """
+
+    logger.info('Computing morph matrix...')
+    subjects_dir = get_subjects_dir(subjects_dir, raise_error=True)
+
+    tris = _get_subject_sphere_tris(subject_from, subjects_dir)
+    maps = read_morph_map(subject_from, subject_to, subjects_dir, xhemi)
+
+    if xhemi:
+        hemi_indexes = [(0, 1), (1, 0)]
+    else:
+        hemi_indexes = [(0, 0), (1, 1)]
+
+    morpher = []
+    for hemi_from, hemi_to in hemi_indexes:
+        idx_use = vertices_from[hemi_from]
+        if len(idx_use) == 0:
+            continue
+        e = mesh_edges(tris[hemi_from])
+        e.data[e.data == 2] = 1
+        n_vertices = e.shape[0]
+        e = e + sparse.eye(n_vertices, n_vertices)
+        m = sparse.eye(len(idx_use), len(idx_use), format='csr')
+        mm = _morph_buffer(m, idx_use, e, smooth, n_vertices,
+                           vertices_to[hemi_to], maps[hemi_from], warn=warn)
+        morpher.append(mm)
+
+    if len(morpher) == 0:
+        raise ValueError("Empty morph-matrix")
+    elif len(morpher) == 1:
+        morpher = morpher[0]
+    else:
+        morpher = sparse_block_diag(morpher, format='csr')
+    logger.info('[done]')
+    return morpher
+
+
+###############################################################################
+# Apply pre-computed morph
+def apply_morph(data_from, morpher, subject_from=None, subject_to=None,
+                **kwargs):
```

Comment:
is the kwargs honestly necessary? not possible to avoid?

Reply:
agree on everything above - for this one: I was wondering the same, but it's a matter of taste / style you want, because there is the problem, that depending on the morph (surf vs vol) I call a function with substantially different **kwargs - so I could either make all kwargs of both functions set with a default in __init__ (messy) or I loop up the kwargs and forward only the necessary onces (those accepted) - or I just forward the entire kwargs list and let the function figure it out (like it is now and that I think is actually quite clean - downside: the user has to know which kwargs are possible...)

Reply:
```
src_surf = forward_surf['src']
stc_surf_orig = ...
source_morph_surf = SourceMorph(src, ...) # (pre-) computes the morph based on if src is for Vol or Surf
stc_surf_morphed = source_morph_surf(stc_surf_orig) # morphs stc
```

Reply:
why not have `VolSourceMorph` and `SurfaceSourceMorph`?

Reply:
I'm gladly redirecting this to @larsoner and @agramfort ? ;)

But I also think that having one single thing doing the morph deciding based on the underlying data makes kinda sense, no?

Reply:
hm... on the other hand: then I could write the file like ``source_estimate.py``, having a ``_BaseMorph`` class and ``VolSourceMorph`` and ``SurfSourceMorph`` inheriting some attributes and load, save, ...